File: georocket-common/src/test/java/io/georocket/util/CoordinateTransformerTest.java
package io.georocket.util;

import org.geotools.referencing.CRS;
import org.geotools.referencing.crs.DefaultGeographicCRS;
import org.junit.Test;
import org.opengis.referencing.FactoryException;
import org.opengis.referencing.crs.CoordinateReferenceSystem;
import org.opengis.referencing.operation.TransformException;

import java.util.Arrays;

import static org.junit.Assert.assertTrue;

/**
 * Test the {@link CoordinateTransformer}
 * @author Tim Hellhake
 */
public class CoordinateTransformerTest {
  /**
   * Test transformation with EPSG code
   */
  @Test
  public void testEPSG() throws FactoryException, TransformException {
    CoordinateTransformer transformer = new CoordinateTransformer("EPSG:31467");
    double[] source = {3477534.683, 5605739.857};
    double[] destination = {8.681739535269804, 50.58691850210496};
    testTransformation(transformer, source, destination);
  }

  /**
   * Test transformation
   * @param transformer the transformer
   * @param source      the source coordinates
   * @param destination the expected destination coordinates in
   *                    {@link DefaultGeographicCRS#WGS84}
   */
  public void testTransformation(CoordinateTransformer transformer, double[] source,
                                 double[] destination)
    throws TransformException {
    double[] transformed = transformer.transform(source, 2);
    assertTrue(transformed[0] - destination[0] < 0.01);
    assertTrue(transformed[1] - destination[1] < 0.01);
  }
}

File: kotlin/src/test/kotlin/io/georocket/util/CoordinateTransformerTest.kt
package io.georocket.util

import org.geotools.referencing.CRS
import org.junit.Assert
import org.junit.Test

/**
 * Test the [CoordinateTransformer]
 * @author Tim Hellhake
 */
class CoordinateTransformerTest {
  /**
   * Test transformation with EPSG code
   * @throws Exception if something has happened
   */
  @Test
  @Throws(Exception::class)
  fun testEPSG() {
    val crs = CRS.decode("EPSG:31467")
    val transformer = CoordinateTransformer(crs)
    val source = doubleArrayOf(3477534.683, 5605739.857)
    val destination = doubleArrayOf(8.681739535269804, 50.58691850210496)
    testTransformation(transformer, source, destination)
  }

  /**
   * Test transformation with WKT code
   * @throws Exception if something has happened
   */
  @Test
  @Throws(Exception::class)
  fun testWKT() {
    val crs = CRS.decode("EPSG:31467")
    val wktCrs = CoordinateTransformer.decode(crs.toWKT())
    val transformer = CoordinateTransformer(wktCrs)
    val source = doubleArrayOf(3477534.683, 5605739.857)
    val destination = doubleArrayOf(8.681739535269804, 50.58691850210496)
    testTransformation(transformer, source, destination)
  }

  /**
   * Test transformation
   * @param transformer the transformer
   * @param source the source coordinates
   * @param destination the expected destination coordinates in
   * [DefaultGeographicCRS.WGS84]
   * @throws Exception if something has happened
   */
  @Throws(Exception::class)
  fun testTransformation(
    transformer: CoordinateTransformer,
    source: DoubleArray?, destination: DoubleArray
  ) {
    val transformed = transformer.transform(source!!, 2)
    Assert.assertEquals(destination[0], transformed!![0], 0.00001)
    Assert.assertEquals(destination[1], transformed[1], 0.00001)
  }
}
